### ISSUE TITLE
Inhale-exhale assertions unsound if used in functions or predicates

### ISSUE THREAD

**Issue by viper-admin** [viperproject/silicon#271]
> Created by **@mschwerhoff** on 2017-03-01 10:45
> Last updated on 2019-08-28 10:07

Silicon's current way of representing the footprints of functions and predicates is unsound in combination with inhale-exhale *assertions* (expressions, i.e. pure assertions, are handled soundly), as illustrated by the example below.

This problem is described in more detail in [my PhD thesis](http://pm.inf.ethz.ch/publications/getpdf.php?bibname=Own&id=Schwerhoff16.pdf) (cf. Section 3.3, subsection "Representing Partial Heaps as Snapshots").

Until the unsoundness if fixed, programs that use the following feature combinations are rejected by Silicon:

- inhale-exhale assertions in predicate bodies
- inhale-exhale assertions in function preconditions

```text

field f: Int
field g: Int

predicate pair(x: Ref) {
  [acc(x.f) && acc(x.g),
   acc(x.g) && acc(x.f)]
}

method test01(x: Ref) {
  inhale acc(x.f) && x.f == 22
  inhale acc(x.g) && x.g == 3

  fold acc(pair(x))
  unfold acc(pair(x))

  assert x.f == 3 && x.g == 22 // Should fail, but unsoundly holds
}



function getf(x: Ref): Int
  requires [acc(x.f) && acc(x.g),
            acc(x.g) && acc(x.f)]
{ x.f }

method test02(x: Ref) {
  inhale acc(x.f) && x.f == 22
  inhale acc(x.g) && x.g == 3

  assert getf(x) == 3 // Should fail, but unsoundly holds
}
```

**Comment by marcoeilers** [comment]
I don't think this is a Silicon issue, I think it's completely unclear what inhale-exhale-assertions in predicates or in function preconditions could possibly mean, and they should almost certainly just be forbidden there. 

Carbon currently allows both and as a result you can verify some absurd things (see below for an example). I'll make a Silver PR to just forbid this in a consistency check (and we should probably discuss this in a Viper meeting).

```
field f: Int
field g: Int

function fn(x: Ref): Int
  requires [acc(x.f), acc(x.g)]
{
  x.f
}

method m(x: Ref)
  requires acc(x.f) && x.f == 2
  requires acc(x.g) && x.g == 3
{
  var tmp: Int
  tmp := fn(x)   // note: we could do this even if we didn't have permission to x.f, which seems absurd
  assert tmp == 2 
  exhale acc(x.f)
  inhale acc(x.f) && x.f == 4
  var tmp2: Int
  tmp2 := fn(x)
  assert tmp2 == 4
  assert tmp2 == tmp
  assert false // all of this verifies
}
```

**Comment by alexanderjsummers** [comment]
I took a look at the snippet above, but I think it looks somewhat reasonable to me. I don't think paired-assertions are meant to prevent potential unsoundness in an encoding (just as inhales don't): wherever they're used the author intends that Viper (for whatever reason) should use one assertion as an assumption in proofs and the other as a requirement whenever checked. So with this example, the author is defining acc(x.g) as the footprint of the function for checking/framing etc. - if the encoding actually guaranteed that x.f and x.g always changed values at the same program points, I guess this would be sound.
